scvi/data/_anndata.py

```diff
@@ -240,12 +241,98 @@ def setup_anndata(
         categorical_covariate_keys,
         continuous_covariate_keys,
     )
+
     logger.info("Please do not further modify adata until model is trained.")
 
+    _verify_and_correct_data_format(adata, data_registry)
+
     if copy:
         return adata
 
 
+def _set_data_in_registry(adata, data, key):
+    """
+    Sets the data associated with key in adata.uns['_scvi']['data_registry'].keys() to data.
+
+    Note: This is a dangerous method and will change the underlying data of the user's anndata
+    Currently used to make the user's anndata C_CONTIGUOUS and csr if it is dense numpy
+    or sparse respectively.
+
+    Parameters
+    ----------
+    adata
+        anndata object to change data of
+    data
+        data to change to
+    key
+        key in adata.uns['_scvi]['data_registry'].keys() associated with the data
+    """
+    use_raw = adata.uns["_scvi"]["use_raw"]
+    data_loc = adata.uns["_scvi"]["data_registry"][key]
+    attr_name, attr_key = data_loc["attr_name"], data_loc["attr_key"]
+
+    if use_raw is True and attr_name in ["X", "var"]:
+        tmp_adata = adata.raw.to_adata()
+    else:
+        tmp_adata = adata
+    if attr_key == "None":
+        setattr(tmp_adata, attr_name, data)
+
+    elif attr_key != "None":
+        attribute = getattr(tmp_adata, attr_name)
+        if isinstance(attribute, pd.DataFrame):
+            attribute.loc[:, attr_key] = data
+        else:
+            attribute[attr_key] = data
+        setattr(tmp_adata, attr_name, attribute)
+
+    if use_raw is True and attr_name in ["X", "var"]:
+        adata.raw = tmp_adata
+
+
+def _verify_and_correct_data_format(adata, data_registry):
+    """
+    Will make sure that the user's anndata is C_CONTIGUOUS and csr if it is dense numpy or sparse respectively.
+
+    Will iterate through all the keys of data_registry.
+
+    Parameters
+    ----------
+    adata
+        anndata to check
+    data_registry
+        data registry of anndata
+    """
+    keys_to_check = [_CONSTANTS.X_KEY, _CONSTANTS.PROTEIN_EXP_KEY]
+    keys = [key for key in keys_to_check if key in data_registry.keys()]
+
+    for k in keys:
+        data = get_from_registry(adata, k)
+        if isspmatrix(data) and (data.getformat() != "csr"):
+            logger.debug("{} is csc_matrix. Overwriting to csr_matrix.".format(k))
+            data = data.tocsr()
+            _set_data_in_registry(adata, data, k)
+        elif isinstance(data, np.ndarray) and (data.flags["C_CONTIGUOUS"] is False):
+            logger.debug(
+                "{} is not C_CONTIGUOUS. Overwriting to C_CONTIGUOUS.".format(k)
+            )
+            data = np.asarray(data, order="C")
+            _set_data_in_registry(adata, data, k)
+        elif isinstance(data, pd.DataFrame) and (
+            data.values.flags["C_CONTIGUOUS"] is False
```

Comment:
values or to_numpy?

Reply:
seems like values is deprecated. just changed to to_numpy()